♻️Feature/android testing update (#643)
getditto/docs: docs/android/quick-tips/testing.mdx

sidebar_position: 2
---

> To test Ditto using Kotlin, we provide you with some basic convenience functions
> you can use to get started.

Reviewer: ```suggestion
Ensure data consistency, replication, storage, and so on by performing various tests on your Kotlin-developed app integrated with the Ditto SDK.  

To simplify testing setup, Ditto provides open-source convenience functions as well as essential configurations and dependencies that you can use in your Kotlin project to test with different instances and directories. 

For Ditto's open-source testing resources:
* For code you can use in your project to test Ditto, see the getditto > samples > [android] (https://github.com/getditto/samples/tree/master/testing/android/) GitHub repo.  
* For essential files, getditto > samples > [dittotesting] https://github.com/getditto/samples/tree/main/testing/android/app/src/androidTest/java/live/ditto/dittotesting) GitHub repo. 

Following are a set of general guidelines and settings to help you get started testing Ditto in your Android app developed with Kotlin.
```

Reviewer: I'm not sure if I have the resources right. Let's let them know where they are going without having to navigate there first.

> For code that you can use in your project, [see the example on GitHub](https://github.com/getditto/samples/tree/master/testing/android/).

Reviewer: ```suggestion
```

Reviewer: Added this link to the repo to the previous list of resources.

### Create custom directories 

> Each instance of Ditto should use a different directory. You can extend
> `DefaultAndroidDittoDependencies` to provide a custom directory.

Reviewer: ```suggestion
Ensure a controlled and isolated testing environment when running tests by using a different directory for each instantiation of the `ditto` class object. 

If you want to use a custom directory instead of `DefaultAndroidDittoDependencies`, the default directory provided in the testing library, using the `persistenceDirectory` class, specify the path where you want to store data:
```

@Test
fun twoDittos() {
	val ditto1 = getDitto(dependenciesWithCustomDirectory(File(getWorkDir(), "ditto1")))
	val ditto2 = getDitto(dependenciesWithCustomDirectory(File(getWorkDir(), "ditto2")))

	ditto1.startSync()
	ditto2.startSync()

	val coll1 = ditto1.store.collection("cars")
	val coll2 = ditto2.store.collection("cars")
	val docId = coll1.upsert(mapOf(
		"make" to "toyota",
		"mileage" to 160000
	))
	coll2.findById(docId).subscribe()
	coll2.findById(docId).observeLocal { doc, event ->
	if (!event.isInitial) {
			doc?.let {

>     		assertEquals(doc["make"].toString(), "toyota")
>       	ditto1.stopSync()
>       	ditto2.stopSync()

Reviewer 2: Consistent indentation
```suggestion
			assertEquals(doc["make"].toString(), "toyota")
			ditto1.stopSync()
			ditto2.stopSync()
```

> 	val ditto1 = getDitto(dependenciesWithCustomDirectory(File(getWorkDir(), "ditto1")))
> 	val ditto2 = getDitto(dependenciesWithCustomDirectory(File(getWorkDir(), "ditto2")))
> 
> 	ditto1.startSync()
> 	ditto2.startSync()
> 
> 	val coll1 = ditto1.store.collection("cars")
> 	val coll2 = ditto2.store.collection("cars")
> 	val docId = coll1.upsert(mapOf(
> 		"make" to "toyota",
> 		"mileage" to 160000
> 	))
> 	coll2.findById(docId).subscribe()
> 	coll2.findById(docId).observeLocal { doc, event ->
> 	if (!event.isInitial) {
> 			doc?.let {
> 			assertEquals(doc["make"].toString(), "toyota")
> 			ditto1.stopSync()
> 			ditto2.stopSync()
> 			}
> 
> 		}
> 	}

Reviewer 2: Indent using spaces

```suggestion
    val ditto1 = getDitto(dependenciesWithCustomDirectory(File(getWorkDir(), "ditto1")))
    val ditto2 = getDitto(dependenciesWithCustomDirectory(File(getWorkDir(), "ditto2")))

    ditto1.startSync()
    ditto2.startSync()

    val coll1 = ditto1.store.collection("cars")
    val coll2 = ditto2.store.collection("cars")
    val docId = coll1.upsert(mapOf(
        "make" to "toyota",
        "mileage" to 160000
    ))
    coll2.findById(docId).subscribe()
    coll2.findById(docId).observeLocal { doc, event ->
        if (!event.isInitial) {
            doc?.let {
                assertEquals(doc["make"].toString(), "toyota")
                ditto1.stopSync()
                ditto2.stopSync()
            }
        }
    }
```
